### PR TITLE
IISHttpContext.FeatureCollection.Protocol recognizes HTTP/2

### PR DESCRIPTION
**Pull request by gfoidl**
Addresses https://github.com/aspnet/AspNetCore/issues/14139

A difficult change :wink:
Could not find any explicit tests for this -- aren't there any?

### CODE REVIEW

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs [hunk 1]
```diff
@@ -86,30 +86,14 @@ private void ExtraFeatureSet(Type key, object value)
...
-                        _httpProtocolVersion = "HTTP/" + protocol.ToString(2);
-                    }
-                }
-                return _httpProtocolVersion;
-            }
-            set
-            {
-                _httpProtocolVersion = value;
-            }
+                { Major: 2, Minor: 0 } => "HTTP/2",
```
Comment (gfoidl) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs]: I thought about using the same approach as in https://github.com/aspnet/AspNetCore/blob/c74e9efd0e6ea5b4d6737d23b92042772a50a249/src/Servers/HttpSys/src/FeatureContext.cs#L177, but

* this code is much more readable (the major/minor-matrix can be read easily)
* it's an explicit interface property that is lazy initialized, so the perf-difference between and object-comparison and two int comparisons should not show up
* in the case that `HttpVersion` is set to a different `Version`-instance (i.e. set not via `Constants.V2`) `HTTP/2.0` may be returned, so here it is the safer way, as `HTTP/2` will be returned

So I'm tempted to update https://github.com/aspnet/AspNetCore/blob/c74e9efd0e6ea5b4d6737d23b92042772a50a249/src/Servers/HttpSys/src/FeatureContext.cs#L174-L193 to the pattern used here. Should I?

Comment (Tratcher) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs]: Yes, feel free to update HttpSys. Or better yet, move this to an API in the code that HttpSys and IIS share.

Comment (gfoidl) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs]: > API in the code that HttpSys and IIS share

I can't find one that is shared 😢 

An extension method 
```c#
public static string GetHttpProtocolVersion(this Version version);
```
would do the job, but where to put this one?

Comment (Tratcher) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs]: This code is all shared: https://github.com/aspnet/AspNetCore/tree/master/src/Shared/HttpSys

File: src/Shared/HttpSys/Extensions.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.HttpSys.Internal
```
Comment (gfoidl) [inline src/Shared/HttpSys/Extensions.cs]: I'm not really happy with this namespace, as it used now in IIS too. Is there a better name / place for this extension?

Comment (Tratcher) [reply src/Shared/HttpSys/Extensions.cs]: It's fine. IIS depends on HttpSys for all transport level functionality so consuming an HttpSys namespace is expected.

File: src/Shared/HttpSys/Constants.cs [hunk 3]
```diff
@@ -15,8 +15,8 @@ internal static class Constants
         internal const string SchemeDelimiter = "://";
         internal const string DefaultServerAddress = "http://localhost:5000";
 
-        internal static Version V1_0 = new Version(1, 0);
-        internal static Version V1_1 = new Version(1, 1);
-        internal static Version V2 = new Version(2, 0);
+        internal static readonly Version V1_0 = new Version(1, 0);
+        internal static readonly Version V1_1 = new Version(1, 1);
+        internal static readonly Version V2 = new Version(2, 0);
```
Comment (Tratcher) [inline src/Shared/HttpSys/Constants.cs]: Are these still used anymore? If not please remove them.

Comment (gfoidl) [reply src/Shared/HttpSys/Constants.cs]: They are used in https://github.com/aspnet/AspNetCore/blob/0d412c3775adf2da1a32599f0ded897933f8f36c/src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs#L157-L174 and https://github.com/aspnet/AspNetCore/blob/0d412c3775adf2da1a32599f0ded897933f8f36c/src/Servers/HttpSys/src/RequestProcessing/Response.cs#L363-L364

File: src/Shared/HttpSys/Extensions.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+{
+    internal static class Extensions
+    {
+        public static string GetHttpProtocolVersion(this Version version) => version switch
```
Comment (anurse) [inline src/Shared/HttpSys/Extensions.cs]: Ooooh, fancy new C# 8.0 features 😜

**Review by analogrelay — COMMENTED** [review]
LGTM but I'll leave it to @Tratcher to sign-off

File: src/Shared/HttpSys/Extensions.cs (hunk 4 above)
Comment (gfoidl) [reply src/Shared/HttpSys/Extensions.cs]: :dart: it's a better showcase than [the one from the docs](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#property-patterns) 😃

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks

**Comment by JamesNK** [comment]
Unit tests that both Kestrel and IIS has the correct value?

**Comment by gfoidl** [comment]
When I should add tests, please give me some pointers where to put them (aside from the question in the description above). For me this are too many server projects* -- so maybe someone else can add these tests...

\* in the positive sense, but I'm not too familiar with them

**Pull request merged**